### PR TITLE
Implement magazines

### PR DESCRIPTION
**Pull request by kidzegeye**
## Overview

Implement the magazine model, resolvers, and repo functions into the backend

## Changes Made

- Created the Magazine model type
- Created queries:
  - getAllMagazines(...): [Magazine!]!
  - getMagazinesBySemester(...): [Magazine!]!
  - getMagazinesByPublicationSlug(...): [Magazine!]!
  - getMagazinesByPublicationSlugs(...): [Magazine!]!
  - getMagazineByID(...): Magazine
  - getMagazinesByIDs(...): [Magazine!]!
  - getFeaturedMagazines(...): [Magazine!]
- Created mutations:
  - incrementMagazineShoutouts(...): Magazine
- Created repo functions to implement resolvers
- Allow magazines to be filtered for profanity
- Added trendiness field to Magazines and Articles
- Modified trendiness formula

## Test Coverage

Ran the following queries and mutations and validated the outputs:
getAllMagazines(...): [Magazine!]!
getMagazinesBySemester(...): [Magazine!]!
getMagazinesByPublicationSlug(...): [Magazine!]!
getMagazinesByPublicationSlugs(...): [Magazine!]!
getMagazineByID(...): Magazine
getMagazinesByIDs(...): [Magazine!]!
getFeaturedMagazines(...): [Magazine!]
incrementMagazineShoutouts(...): Magazine

## Next Steps (delete if not applicable)

Push magazine download functionality on the microservice

### CODE REVIEW

File: src/repos/MagazineRepo.ts
Comment (chalo2000) [inline src/repos/MagazineRepo.ts:123]: how did we come up with this number? why do we randomly sample? shouldn't be deterministically be showing the new featured magazines rather than getting 100 random magazines from our database and showing those that are new/featured? also is there a difference between a new magazine and a featured magazine? if not then `Get new featured magazines` is misleading, just put `Get featured magazines`

File: src/repos/MagazineRepo.ts
Comment (chalo2000) [inline src/repos/MagazineRepo.ts:134]: why did we disable the linter for this line? if possible, we should disable a specific lint so it's clearer why it was disabled in the first place

File: src/resolvers/ArticleResolver.ts
Comment (chalo2000) [inline src/resolvers/ArticleResolver.ts:114]: this seems unrelated to this PR, why did we have to divide this by 1000

File: src/resolvers/MagazineResolver.ts
Comment (chalo2000) [inline src/resolvers/MagazineResolver.ts:24]: nice comment

File: src/resolvers/MagazineResolver.ts
Comment (chalo2000) [inline src/resolvers/MagazineResolver.ts:89]: but then here we don't divide by 1000? let's standardize this formula because I didn't even know how trendiness was calculated

**Review by chalo2000 — CHANGES_REQUESTED** [review]
noice, left some comments that should be addressed before merging

File: src/repos/MagazineRepo.ts
Comment (kidzegeye) [reply src/repos/MagazineRepo.ts:42]: Forgot to add that to the model, its a field in the mongo schema

File: src/resolvers/ArticleResolver.ts
Comment (kidzegeye) [reply src/resolvers/ArticleResolver.ts:114]: The trendiness value was very small otherwise, as the dates in the denominator were in milliseconds

Comment (kidzegeye) [reply src/resolvers/ArticleResolver.ts:114]: It was to keep it consistent with the magazine trendiness formula

File: src/resolvers/MagazineResolver.ts
Comment (kidzegeye) [reply src/resolvers/MagazineResolver.ts]: fixed

File: src/repos/MagazineRepo.ts
Comment (kidzegeye) [reply src/repos/MagazineRepo.ts:134]: The linter had a rule against dangling underscores, added a file-wide disable for  no-underscore-dangle

File: src/repos/MagazineRepo.ts
Comment (kidzegeye) [reply src/repos/MagazineRepo.ts:124]: This is to be consistent with how trending articles are chosen. We are thinking about using a "trendiness" value that was calculated from shoutouts that has gone unused so far, but I would like to do that in a different PR for both magazines and articles

File: src/app.ts
Comment (chalo2000) [inline src/app.ts:46]: has this change been communicated with frontend? changing this will break current apps fetching articles

File: src/repos/MagazineRepo.ts
Comment (chalo2000) [inline src/repos/MagazineRepo.ts:14]: when would we disable this filter? shouldn't this be an environment variable so that we can flip it on and off without having to make a new push?

**Review by chalo2000 — APPROVED** [review]
LGTM

File: src/app.ts
Comment (kidzegeye) [reply src/app.ts:46]: I'm not sure if frontend uses these endpoints

File: src/repos/MagazineRepo.ts
Comment (kidzegeye) [reply src/repos/MagazineRepo.ts:14]: We want to disable it now because there isn't any covid content policy on the apple store anymore

Comment (kidzegeye) [reply src/repos/MagazineRepo.ts:14]: It was disabled in another pr, but I can change it to an env variable there

**Review by Archit404Error — APPROVED** [review]
Good to go once we look into the /collect/ endpoint usage on frontend

**Comment by tedim52** [comment]
can't wait to see this feature come out!

**Pull request merged**
